pkg/clockcache/cache.go

```diff
@@ -40,9 +42,13 @@ type element struct {
 	key   interface{}
 	value interface{}
 
-	// CLOCK marker if this is recently used
-	used uint32
-	size uint64
+	// CLOCK marker if this element is recently used together with timestamp
+	// To avoid memory waist we pack together `used` flag with timestamp into 32 bits
+	// Left most bit represents if element was used (1 or 0). The rest (31 bits) represent timestamp
+	// Timestamp is helping us figure out if we should attempt to grow cache or not
+	// Timestamp is stored in Unix seconds (31 bits should be enough for the next 15 years)
+	usedWithTs uint32
```

Comment:
Can we take from the `_ [16]bytes` that we are using for padding and store the TS as a separate int64?

Reply:
I don't think that's possible. If I am not mistaken `[16]bytes` are there to guarantee alignment of 16 bytes.

Reply:
So padding was added to prevent cache-line bouncing that's (potentially) happening on concurrent access. I've just run local benchmark of `BenchmarkCacheFalseSharing` and there was no difference with and without padding. 😕  I think we should consider removing [16] bytes of padding if it provides no CPU gain since it clearly waists memory... I'd leave that out of the scope of this PR but happy to revisit it.

Reply:
Fine by me

Reply:
I think that it should be possible to "steal" this field from the padding bytes, as @alejandrodnm suggested.

> I've just run local benchmark of BenchmarkCacheFalseSharing and there was no difference with and without padding.

What platform are you on? I noticed that on M1 macs the cache alignment doesn't make a difference, but it does make a difference on x86 platforms.

Reply:
I was testing on M1 and made no difference although the comment stated that there is about 16% difference on Mac M1?! Ideally we could remove padding and save a lots of memory.

>I noticed that on M1 macs the cache alignment doesn't make a difference, but it does make a difference on x86 platforms.

How much is the difference there?

Reply:
>I think that it should be possible to "steal" this field from the padding bytes, as @alejandrodnm suggested.

Can you explain your reasoning.  Why would we pad on smaller amount of bytes (I assumed there was a reason we wanted to pad on 16 bytes, otherwise it's padded on 8 bytes b/c of int64?

Reply:
I've just run bench tests on x86. It seems difference between padding and no padding is statistically insignificant. Not sure how did you get 2X difference 😕 

```
root@my-bench-test:/promscale/pkg/clockcache# cat no-padding.txt
goos: linux
goarch: amd64
pkg: github.com/timescale/promscale/pkg/clockcache
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
BenchmarkCacheFalseSharing-2   	 2510319	       490.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2503368	       489.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2464582	       494.0 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2555012	       482.5 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2532994	       470.0 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2496974	       466.4 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2564890	       471.7 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2497118	       482.9 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2585713	       477.0 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2540541	       470.8 ns/op	      15 B/op	       1 allocs/op
PASS
ok  	github.com/timescale/promscale/pkg/clockcache	16.974s
```

And with padding:
```
root@my-bench-test:/promscale/pkg/clockcache# cat padding.txt
goos: linux
goarch: amd64
pkg: github.com/timescale/promscale/pkg/clockcache
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
BenchmarkCacheFalseSharing-2   	 2926026	       408.5 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2917959	       428.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2709094	       423.8 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2362786	       525.4 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2888036	       493.4 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2351191	       507.8 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2457351	       516.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2332888	       498.7 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2971623	       432.9 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2854483	       515.7 ns/op	      15 B/op	       1 allocs/op
PASS
ok  	github.com/timescale/promscale/pkg/clockcache	19.381s
```
I ran: `go test -bench=BenchmarkCacheFalseSharing -cpu=2 -count=10 -run=^$`
I was running tests in K8s environment with 4 CPUs. However I would be really surprised if we would get 2X on bare metal - I can verify it once I get my hands on the old laptop.
I did repeat my runs couple of time without major difference.

Reply:
Hmm... I'm obviously misremembering that 2x.

I just tested this out on some AWS VMs as well. Maybe the 2x that I had in mind was the difference between how long each operation took on Intel vs. M1 (~200ns vs ~400ns).

Anyway, here are the results on an AMD x86_64 system:
```
ubuntu@ip-172-31-17-88:~/promscale/pkg/clockcache$ go test -bench=BenchmarkCacheFalseSharing -cpu=2 -count=10 | tee padding.txt
goos: linux
goarch: amd64
pkg: github.com/timescale/promscale/pkg/clockcache
cpu: AMD EPYC 7R13 Processor
BenchmarkCacheFalseSharing-2   	 5070142	       240.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5058939	       241.7 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5113359	       224.2 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5607207	       238.7 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4946768	       219.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5392538	       240.5 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4942778	       226.3 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5245922	       241.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5132740	       217.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 5445062	       236.0 ns/op	      15 B/op	       1 allocs/op
PASS
ok  	github.com/timescale/promscale/pkg/clockcache	20.516s

ubuntu@ip-172-31-17-88:~/promscale/pkg/clockcache$ go test -bench=BenchmarkCacheFalseSharing -cpu=2 -count=10 | tee no-padding.txt
goos: linux
goarch: amd64
pkg: github.com/timescale/promscale/pkg/clockcache
cpu: AMD EPYC 7R13 Processor
BenchmarkCacheFalseSharing-2   	 4700396	       251.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4702579	       252.2 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4681315	       253.8 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4787164	       253.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4723281	       253.2 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4739452	       252.5 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4748074	       254.3 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4760126	       255.0 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4706127	       253.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 4706882	       254.5 ns/op	      15 B/op	       1 allocs/op
PASS
ok  	github.com/timescale/promscale/pkg/clockcache	20.693s

ubuntu@ip-172-31-17-88:~/promscale/pkg/clockcache$ ~/go/bin/benchstat no-padding.txt padding.txt
name                 old time/op    new time/op    delta
CacheFalseSharing-2     253ns ± 1%     233ns ± 6%  -8.21%  (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
CacheFalseSharing-2     15.0B ± 0%     15.0B ± 0%    ~     (all equal)

name                 old allocs/op  new allocs/op  delta
CacheFalseSharing-2      1.00 ± 0%      1.00 ± 0%    ~     (all equal)
```

And on an Intel x86_64 system:

```
ubuntu@ip-172-31-24-242:~/promscale/pkg/clockcache$ go test -bench=BenchmarkCacheFalseSharing -cpu=2 -count=10 | tee padding.txt
goos: linux
goarch: amd64
pkg: github.com/timescale/promscale/pkg/clockcache
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
BenchmarkCacheFalseSharing-2   	 2494210	       411.9 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2985262	       492.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 3059500	       408.8 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2997885	       412.2 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2947947	       423.0 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 3034550	       413.4 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 3029181	       414.5 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 3026805	       433.5 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 3033912	       442.4 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2942834	       421.6 ns/op	      15 B/op	       1 allocs/op
PASS
ok  	github.com/timescale/promscale/pkg/clockcache	34.228s

ubuntu@ip-172-31-24-242:~/promscale/pkg/clockcache$ go test -bench=BenchmarkCacheFalseSharing -cpu=2 -count=10 | tee no-padding.txt
goos: linux
goarch: amd64
pkg: github.com/timescale/promscale/pkg/clockcache
cpu: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
BenchmarkCacheFalseSharing-2   	ben 2200164	       473.0 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2439961	       453.4 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2496398	       496.9 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2463030	       488.6 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2362767	       491.9 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2445717	       497.1 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2410858	       488.7 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2452435	       527.7 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2280991	       492.3 ns/op	      15 B/op	       1 allocs/op
BenchmarkCacheFalseSharing-2   	 2286775	       475.9 ns/op	      15 B/op	       1 allocs/op
PASS
ok  	github.com/timescale/promscale/pkg/clockcache	26.548s

ubuntu@ip-172-31-24-242:~/promscale/pkg/clockcache$ ~/go/bin/benchstat no-padding.txt padding.txt
name                 old time/op    new time/op    delta
CacheFalseSharing-2     489ns ± 8%     420ns ± 5%  -14.00%  (p=0.000 n=10+9)

name                 old alloc/op   new alloc/op   delta
CacheFalseSharing-2     15.0B ± 0%     15.0B ± 0%     ~     (all equal)

name                 old allocs/op  new allocs/op  delta
CacheFalseSharing-2      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

Note that the following diff between the `padding` and `no-padding` runs:
```diff
ubuntu@ip-172-31-24-242:~/promscale/pkg/clockcache$ git diff
diff --git a/pkg/clockcache/cache.go b/pkg/clockcache/cache.go
index 43448ca7..f57974c5 100644
--- a/pkg/clockcache/cache.go
+++ b/pkg/clockcache/cache.go
@@ -45,7 +45,7 @@ type element struct {
        size uint64

        // pad Elements out to be cache-aligned, see BenchmarkCacheFalseSharing
-       _ [16]byte
+       //_ [16]byte
 }

 func WithMax(max uint64) *Cache {
diff --git a/pkg/clockcache/cache_test.go b/pkg/clockcache/cache_test.go
index e65f301d..59a99342 100644
--- a/pkg/clockcache/cache_test.go
+++ b/pkg/clockcache/cache_test.go
@@ -314,6 +314,7 @@ func TestReset(t *testing.T) {
 }

 func TestElementCacheAligned(t *testing.T) {
+       t.Skip()
        elementSize := unsafe.Sizeof(element{})
        if elementSize%64 != 0 {
                t.Errorf("unaligned element size: %d", elementSize)
```

Anyway, needless to say there does seem to be a difference. But, is it worth sweating over? Honestly, probably not.

Reply:
In some of my runs I did see around 5% diff (for padded version), but I feel the same: probably not worth the `cost`. On some average case with around 5M series we end up spending close to 100MB of memory... - only for series cache....
Anyways let's leave this for a separate PR, but I am in favor of removing extra 16 bytes of padding....